lib/board.rb

```diff
@@ -30,4 +29,35 @@ def cells
   def valid_coordinate?(coordinate)
     @cells.key?(coordinate)
   end
+
+  def valid_placement?(ship, coordinates)
+    if (ship.length == coordinates.length) &&
+       (coordinates.all? { |coordinate| valid_coordinate?(coordinate) }) &&
+        !coordinates.empty?
+        consecutive_coordinates?(coordinates)
+    else
+      false
+    end
+  end
+
+  def consecutive_coordinates?(coordinates)
+    letters = coordinates.map { |coordinate| coordinate[0] }
+    numbers = coordinates.map { |coordinate| coordinate[1].to_i }
+
+    if !consecutive_letters?(letters) && consecutive_numbers?(numbers)
+      true
+    elsif consecutive_letters?(letters) && !consecutive_numbers?(numbers)
+      true
+    else 
+      false
+    end
+  end
+
+  def consecutive_numbers?(numbers)
+    numbers.each_cons(2).all? { |first, second| second - first == 1 }
```

Comment:
Could you explain each_cons method to me, as the ruby docs does not go into much detail? Though I have done some research on it, and have a basic understanding  i would like to know more in detail.

Reply:
Each-cons is an enumerable that takes an argument (group size). It slides through a collection 1 element at a time, starting from the first element, and ending when it hits the final element.
For example:
array = [1, 2, 3, 4, 5]
array.each_cons(2) { |num1, num2| puts "#{num1} and #{num2}" }
The output would look like this:
1 and 2
2 and 3
3 and 4
4 and 5

By using this method we can check that each element is consecutive to the previous element by using the .all? method, subtraction, and a comparison operator. 
In our case, we are validating the consecutiveness of coordinates!

Reply:
That makes sense! Thank you for the clarification!